docs/spec/spec.md

```diff
@@ -0,0 +1,395 @@
+
+This design supports the following cardinalities:
+1. One-to-one (1-1)
+2. One-to-many (1-n)
+
+Relationship is mandatory in both entities.
+
+#### 2.3.1. One-to-one (1-1)
+ 
+1-1 relationship is defined by a field of type `EntityType` in both entities. 
+
+Example: Each User must have a Car, and each Car must have an owner.
+
+```ballerina
+type Car record {|
+   readonly int id;
+   string name;
+   User owner;
+|};
+
+type User record {|
+   readonly int id;
+   string name;
+   Car car;
+|};
+```
+
+The first record `Car` is taken as the parent in the 1-1 relationship and will contain the foreign key of the second record `User`.
+
+The default foreign key field name will be `userId` in the `Car` table, which refers to the identifier field of the `User` table by default. (`<lowercasedAssociatedEntityName><First-Letter Capitalized IdentifierFieldName>`)
+
+#### 2.3.2. One-to-Many (1-n)
+
+1-n relationship is defined by a field of type `EntityType` in one entity and `EntityType[]` in the other.
+
+Example: A User can own zero or more cars, and a Car must have an owner. 
+
+```ballerina
+type Car record {|
+   readonly int id;
+   string name;
+   User owner;
+|};
+
+type User record {|
+   int id;
+   string name;
+   Car[] cars;
+|};
+```
+The entity that contains the field of type `EntityType` is taken as the parent in the 1-n relationship and will contain the foreign key of the other entity.
+
+The default foreign key field name will be `userId` in the `Car` table, which refers to the identifier field of the `User` table by default. (`<lowercasedAssociatedEntityName><First-Letter Capitalized IdentifierFieldName>`)
+
+## 3. Derived Entity Types and Persist Clients
+
+Derived Entity Types and Clients can be easily generated from Persist CLI tool.
+
+### 3.1. Derived Entity Types
+
+Ballerina record types used in the `PersistClient` are derived from the entity types defined in the data model.
+
+Example of a defined entity,
+```ballerina
+type Workspace record {|
+    readonly string workspaceId;
+    string workspaceType;
+    Building location; // 1-n relation (parent)
+    Employee employee; // 1-1 relation (child)
+|};
+```
+
+There are 3 types of derived entity types:
+1. Entity Types  
+    These are the types defined in the data model. These are used to define the structure of the data in the data source. 
+    The entity who is the parent of the association will include foreign key field.
+    ```ballerina
+    public type Workspace record {|
+        readonly string workspaceId;
+        string workspaceType;
+        string buildingBuildingCode;
+    |};
+    ```
+    
+2. Insert Types
+    These are records used to insert data in the data source. It is the same as the entity type.
+    ```ballerina
+    type WorkspaceInsert Workspace;
+    ```
+
+3. Update Types
+    These are records used to update data in the data source. These are entity types without the identifier fields.
+    ```ballerina
+    type WorkspaceUpdate record {|
```

Comment:
shouldn't these be nullable types, as in [here](https://github.com/ballerina-platform/module-ballerina-persist/blob/005e6fafd3a21fec030880da856a604728eac401/ballerina/tests/generated_types.bal#L58)

Reply:
What you mean is its optional field? `int field?;` ?

Reply:
Sorry it should be `optional field`